crypto/cms/cms_ess.c

```diff
@@ -335,3 +337,87 @@ ASN1_OCTET_STRING *cms_encode_Receipt(CMS_SignerInfo *si)
     CMS_ReceiptRequest_free(rr);
     return os;
 }
+
+/*
+ * Add signer certificate's V2 digest to a SignerInfo
+ * structure
+ */
+
+CMS_SignerInfo *CMS_add1_signing_cert_v2(CMS_SignerInfo *si, 
+                                         const EVP_MD *sign_md)
+{
+    ASN1_STRING *seq = NULL;
+    unsigned char *p, *pp = NULL;
+    ESS_SIGNING_CERT_V2 *sc = NULL;
+    int len, r = 0;
+
+    if ((sc = ESS_SIGNING_CERT_V2_new_init(sign_md, si->signer, NULL, 1)) == NULL)
+        goto err;
+    
+    /* Add SigningCertificateV2 signed attribute to the signer info. */
+    len = i2d_ESS_SIGNING_CERT_V2(sc, NULL);
+    if ((pp = OPENSSL_malloc(len)) == NULL)
+        goto err;
+    p = pp;
+    i2d_ESS_SIGNING_CERT_V2(sc, &p);
+    if (!(seq = ASN1_STRING_new()) || !ASN1_STRING_set(seq, pp, len))
+        goto err;
+    OPENSSL_free(pp);
+    ESS_SIGNING_CERT_V2_free(sc);
+    pp = NULL;
+    if (!CMS_signed_add1_attr_by_NID(si, NID_id_smime_aa_signingCertificateV2,
+                                     V_ASN1_SEQUENCE, seq, -1))
+        goto err;
+    r = 1;
+
+ err:
+    ASN1_STRING_free(seq);
+    if (r)
+        return si;
+    CMSerr(CMS_F_CMS_ADD1_SIGNING_CERT_V2, ERR_R_MALLOC_FAILURE);
+    OPENSSL_free(pp);
+    ESS_SIGNING_CERT_V2_free(sc);
+    return NULL;
+}
+
+/*
+ * Add signer certificate's digest to a SignerInfo
+ * structure
+ */
+
+CMS_SignerInfo *CMS_add1_signing_cert(CMS_SignerInfo *si)
+{
+    ASN1_STRING *seq = NULL;
+    unsigned char *p, *pp = NULL;
+    ESS_SIGNING_CERT *sc = NULL;
+    int len, r = 0;
+
+    if ((sc = ESS_SIGNING_CERT_new_init(si->signer, NULL, 1)) == NULL)
```

Comment:
There a big API misunderstanding here : 
SigningCertificate is tie to SignerInfo, so it becomes a kind of duplicate self-reference.

You should create outside SignerInfo, even reusing the same variable as for SignerIfno, and them add it to SignerInfo.

Reply:
So this method : `int CMS_add1_signing_cert(CMS_SignerInfo *si, ESS_SIGNING_CERT *sc)` 
 '